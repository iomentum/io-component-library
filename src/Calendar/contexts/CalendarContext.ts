import { createContext, Dispatch, SetStateAction } from "react";
import { Display, MomentRangeExtended } from "../utils";

export interface CalendarContextInterface {
  display: Display;
  setDisplay: Dispatch<SetStateAction<Display>>;
  setDate: Dispatch<SetStateAction<MomentRangeExtended>>;
}

export const CalendarContext = createContext<CalendarContextInterface>(null);
