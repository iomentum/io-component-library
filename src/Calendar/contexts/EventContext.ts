import { createContext, Dispatch, SetStateAction } from 'react';
import { EventAction, Event } from '../reducers/EventReducer';

export interface EventContextInterface {
  eventsCollection: Event[];
  setEventsCollection: Dispatch<SetStateAction<Event[]>>;
  eventManagementOpened: boolean;
  setEventManagementOpened: Dispatch<SetStateAction<boolean>>;
  event: Event;
  dispatchEvent: Dispatch<EventAction>;
}

export const EventContext = createContext<EventContextInterface>(null);
