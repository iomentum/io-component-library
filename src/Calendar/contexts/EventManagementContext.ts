import { createContext, Dispatch, SetStateAction } from 'react';
import { EventAction, EventModel } from '../reducers/EventReducer';

export interface EventManagementContext {
  fullDayEvent: boolean;
  event: EventModel;
  setFullDayEvent: Dispatch<SetStateAction<boolean>>;
  dispatchEvent: Dispatch<EventAction>;
}

export const EventManagementContext = createContext<EventManagementContext>(null);
