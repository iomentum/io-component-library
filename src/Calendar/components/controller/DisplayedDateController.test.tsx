/**
 * @jest-environment jsdom
 */

import { render, screen } from '@testing-library/react';
import React from 'react';
import { CalendarContext } from '../../contexts/CalendarContext';
import {
  computeNewDate,
  ControlButton,
  DisplayedDateController,
  MathOperation,
} from './DisplayedDateController';

const calendarContextMock = (component, { providerValue, ...renderOptions }) =>
  render(
    <CalendarContext.Provider value={providerValue}>{component}</CalendarContext.Provider>,
    renderOptions
  );

describe('ControlButton component', () => {
  describe('@snapshot', () => {
    it('should match with previous ControlButton', () => {
      const { asFragment } = render(<ControlButton label="test" onClick={jest.fn()} />);

      expect(asFragment()).toMatchSnapshot('ControlButton snapshot');
    });
  });

  describe('@event', () => {
    it('should handle a click', () => {
      const handleOnClick = jest.fn();
      render(<ControlButton label="" onClick={handleOnClick} />);

      screen.getByRole('button').click();

      expect(handleOnClick).toHaveBeenCalledTimes(1);
    });
  });
});

describe('DisplayedDateController component', () => {
  describe('@snapshot', () => {
    it('should match with previous ', () => {
      const providerValue = {
        setDisplayedDate: jest.fn(),
      };
      const { asFragment } = calendarContextMock(<DisplayedDateController />, {
        providerValue,
      });

      expect(asFragment()).toMatchSnapshot('DisplayedDateController snapshot');
    });
  });

  describe('@event', () => {
    it('each buttons should trigger a setDisplayedDate on click', () => {
      const providerValue = {
        setDisplayedDate: jest.fn(),
      };
      calendarContextMock(<DisplayedDateController />, { providerValue });

      screen.getAllByRole('button').forEach((button) => button.click());

      expect(providerValue.setDisplayedDate).toHaveBeenCalledTimes(3);
    });
  });
});

describe('computeNewDate function', () => {
  describe('should add', () => {
    it('1 day', () => {
      const date = new Date();
      const computedDate = computeNewDate(new Date(date), MathOperation.Addition, 1);

      date.setDate(date.getDate() + 1);
      expect(computedDate).toEqual(new Date(date));
    });

    it('7 day', () => {
      const date = new Date();
      const computedDate = computeNewDate(new Date(date), MathOperation.Addition, 7);

      date.setDate(date.getDate() + 7);
      expect(computedDate).toEqual(new Date(date));
    });
  });

  describe('should substract', () => {
    it('1 day', () => {
      const date = new Date();
      const computedDate = computeNewDate(new Date(date), MathOperation.Substraction, 1);

      date.setDate(date.getDate() - 1);
      expect(computedDate).toEqual(new Date(date));
    });

    it('7 day', () => {
      const date = new Date();
      const computedDate = computeNewDate(new Date(date), MathOperation.Substraction, 7);

      date.setDate(date.getDate() - 7);
      expect(computedDate).toEqual(new Date(date));
    });
  });
});
