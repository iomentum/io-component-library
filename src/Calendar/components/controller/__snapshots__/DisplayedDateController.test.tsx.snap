// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlButton component @snapshots should match with previous ControlButton: ControlButton snapshot 1`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      test
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;

exports[`DisplayedDateController component @snapshots should match with previous : DisplayedDateController snapshot 1`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      &lt;
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Today
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      &gt;
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;
